pkg/text/font.go

```diff
@@ -11,51 +11,121 @@ type Font struct {
 	tx *graphics.Texture
 }
 
+type FontProps struct {
+	StrokeWidth float32
+	StrokeEdge float32
+	Color mgl32.Vec4
+}
+
+var (
+	FontPropLarge = FontProps{0.8, 0.02, mgl32.Vec4{1,1,1,1}}
+	FontPropSmall = FontProps{0.5, 0.1, mgl32.Vec4{1,1,1,1}}
+)
+
 func NewFontFromFiles(bmpath, texpath string) *Font {
 	f := &Font{}
 	f.bm = NewBmFontFromFile(bmpath)
 	f.tx = graphics.NewTextureFromFile(texpath)
 	return f
 }
 
-func (f *Font) RenderText(txt string) *graphics.Primitive2D {
-	vertices := make([]float32, len(txt)*8+(len(txt)-1)*4)
-	uvCoords := make([]float32, len(txt)*8+(len(txt)-1)*4)
-
-	idx := 0
-	end := len(txt)
-	for ic, char := range txt {
-		if bmc, ok := f.bm.Characters[string(char)]; ok {
-			log.Printf("%+v", bmc)
-			copy(vertices[idx:], []float32{
-				float32(ic),1,
-				float32(ic+1),1,
-				float32(ic),0,
-				float32(ic+1),0,
-			})
-			copy(uvCoords[idx:], []float32{0,1,1,1,0,0,1,0,})
-			idx += 8
-
-			if ic < end {
-				copy(vertices[idx:], []float32{1,1,1,1})
-				copy(uvCoords[idx:], []float32{1,1,1,1})
-				idx += 4
-			}
-		} else {
+const CharVertices = 12
+
+func charQuad(offsetX, offsetY, width, height float32) []float32 {
+	q := [CharVertices]float32{
+		offsetX, offsetY+height,       // bl
+		offsetX+width, offsetY+height, // br
+		offsetX, offsetY,              // tl
+		offsetX, offsetY,              // tl
+		offsetX+width, offsetY+height, // br
+		offsetX+width, offsetY,        // tr
+	}
+	return q[:]
+}
+
+func (f *Font) RenderText(
+	txt string,
+	position mgl32.Vec3,
+	size mgl32.Vec2,
+	fp FontProps,
+) *graphics.Primitive2D {
+	var (
+		vnum = len(txt)*CharVertices
+		vertices = make([]float32, vnum)
+		uvCoords = make([]float32, vnum)
+		idx = 0
+		cursorX float32 = 0
+		cursorY float32 = 0
+		scale = 1.0/float32(f.bm.lineHeight)
+		uvscalex = 1.0/float32(f.bm.pageWidth)
+		uvscaley = 1.0/float32(f.bm.pageHeight)
+	)
+
+	for _, char := range txt {
+		if char == 0x0a {
+			cursorX = 0
+			cursorY += 1
+			continue
+		}
+		bmc, ok := f.bm.Characters[string(char)]
+		if !ok {
 			log.Printf(
 				"ERR: char %v (%v) not found in font map",
 				string(char), char,
 			)
+			continue
 		}
+		copy(
+			vertices[idx:],
+			charQuad(
+				cursorX+float32(bmc.offsetX)*scale,
+				cursorY+float32(bmc.offsetY)*scale,
+				float32(bmc.width)*scale,
+				float32(bmc.height)*scale,
+			),
+		)
+		copy(
+			uvCoords[idx:],
+			charQuad(
+				float32(bmc.x)*uvscalex,
+				float32(bmc.y)*uvscaley,
+				float32(bmc.width)*uvscalex,
+				float32(bmc.height)*uvscaley,
+			),
+		)
+		idx += CharVertices
+		cursorX += float32(bmc.advanceX)*scale
 	}
 
-	p := graphics.NewTriangleStrip(
-		vertices,
-		uvCoords,
-		f.tx,
-		mgl32.Vec3{0,0,0},
-		mgl32.Vec2{80,80},
+	shaderProgram := graphics.NewShaderProgram(
+		graphics.VertexShaderPrimitive2D,
+		"",
+		FragmentDistanceFieldFont,
 	)
-	p.SetAnchorToCenter()
-	return p
+
+	// TODO: this should be done in draw/drawInBatch?
```

Comment:
Not sure about this question. In the Draw() method you need to setup everything, in the drawInBatch() you already have texture and shader setup, you just need to send the uniforms

Reply:
yes, i'll need to make a separate primitive type and override draw/drawinbatches